clients/src/main/java/org/apache/kafka/clients/consumer/NoOffsetForPartitionException.java

```diff
@@ -26,19 +26,20 @@
 
     private static final long serialVersionUID = 1L;
 
-    private final TopicPartition partition;
+    private final Set<TopicPartition> partitionSet;
 
-    public NoOffsetForPartitionException(TopicPartition partition) {
-        super("Undefined offset with no reset policy for partition: " + partition);
-        this.partition = partition;
+    public NoOffsetForPartitionException(String message) {
```

Comment:
Agreed, some convention or improvement is needed. 

The problem here is that [Errors.java](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/protocol/Errors.java) needs to be able to throw exceptions based on an error code alone (based on the pattern defined by all other errors).

I left the constructor that takes a partition for compatibility, though internally the partitions() and partition() methods are not used. In fact, currently, having access to the partition may not be all that useful. The exception is thrown from [resetOffset()](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java#L280) but that is called by [updateFetchPositions()](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java#L155). Even though, updateFetchPositions takes multiple TopicPartitions, only  the first exception will be thrown, and all others in error won't be known. So, regardless you will need to check/fix/handle all TopicPartitions requested.

I am not a huge fan of using exceptions to pass information in unexceptional use cases. If it is expected to need to handle issues like not having an offset for a given partition regularly in basic client code, I would rather see that built into the api than hidden inside exceptions. This also avoids issue like above, where only 1 exception is/can be thrown. Instead its a part of the request/response contract to handle any issues (or ignore them at your own risk).